Menu update

Hai-tun opened the pull request:
Upgrade menu module

Commit:
Separate cancel top or bottom click

Commit:
Strengthen anti illegal top drag

Commit:
cleanup code

Commit:
Add anti illegal bottom inventory double click

Commit:
Add anti illegal bottom inventory shift click

Commit:
Change cancelOnBottomClick default value to false

DevSrSouza commented:
Hey @Hai-tun, this for the PR, but, you can explain the changes that this PR does?

Hai-tun commented:
***
`menu` : Separate `cancelOnClick` into `cancelOnTopClick(defalut = true)` and `cancelOnBottomClick(defalut = false)`
`(The bottom inventory will be clickable and moveable)`
***
`top drag` : Changed from canceling the event to returning the illegally dragged item back to the cursor
***
`bottom inventory double click` : Add anti illegal bottom inventory shift click
***
`bottom inventory shift click` : Add anti illegal bottom inventory shift click
***

Commit:
Fix genericExecutor higher priority than playerExecutor
related errors:
[13:49:45] [Server thread/WARN]: java.lang.ClassCastException: org.bukkit.craftbukkit.v1_16_R3.command.ColouredConsoleSender cannot be cast to org.bukkit.entity.Player
[13:49:45] [Server thread/WARN]: 	at moe.orangemc.userservice.Main$onPluginEnable$5$1.invokeSuspend(Main.kt:83)
[13:49:45] [Server thread/WARN]: 	at moe.orangemc.userservice.Main$onPluginEnable$5$1.invoke(Main.kt)
[13:49:45] [Server thread/WARN]: 	at br.com.devsrsouza.kotlinbukkitapi.dsl.command.CommandDSL$execute$1$1.invokeSuspend(CommandDSL.kt:193)
[13:49:45] [Server thread/WARN]: 	at br.com.devsrsouza.kotlinbukkitapi.dsl.command.CommandDSL$execute$1$1.invoke(CommandDSL.kt)
[13:49:45] [Server thread/WARN]: 	at br.com.devsrsouza.kotlinbukkitapi.dsl.command.CommandDSL.treatFail(CommandDSL.kt:251)
[13:49:45] [Server thread/WARN]: 	at br.com.devsrsouza.kotlinbukkitapi.dsl.command.CommandDSL$execute$1.invokeSuspend(CommandDSL.kt:192)
[13:49:45] [Server thread/WARN]: 	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:56)
[13:49:45] [Server thread/WARN]: 	at com.okkero.skedule.BukkitDispatcher.dispatch(BukkitDispatcher.kt:45)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.DispatchedContinuationKt.resumeCancellableWith(DispatchedContinuation.kt:288)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.intrinsics.CancellableKt.startCoroutineCancellable(Cancellable.kt:26)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.CoroutineStart.invoke(CoroutineStart.kt:109)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.AbstractCoroutine.start(AbstractCoroutine.kt:158)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch(Builders.common.kt:56)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.BuildersKt.launch(Unknown Source)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.BuildersKt__Builders_commonKt.launch$default(Builders.common.kt:49)
[13:49:45] [Server thread/WARN]: 	at kotlinx.coroutines.BuildersKt.launch$default(Unknown Source)
[13:49:45] [Server thread/WARN]: 	at br.com.devsrsouza.kotlinbukkitapi.dsl.command.CommandDSL.execute(CommandDSL.kt:190)
[13:49:45] [Server thread/WARN]: 	at org.bukkit.command.SimpleCommandMap.dispatch(SimpleCommandMap.java:149)
[13:49:45] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_16_R3.CraftServer.dispatchCommand(CraftServer.java:763)
[13:49:45] [Server thread/WARN]: 	at org.bukkit.craftbukkit.v1_16_R3.CraftServer.dispatchServerCommand(CraftServer.java:748)
[13:49:45] [Server thread/WARN]: 	at net.minecraft.server.v1_16_R3.DedicatedServer.handleCommandQueue(DedicatedServer.java:387)
[13:49:45] [Server thread/WARN]: 	at net.minecraft.server.v1_16_R3.DedicatedServer.b(DedicatedServer.java:356)
[13:49:45] [Server thread/WARN]: 	at net.minecraft.server.v1_16_R3.MinecraftServer.a(MinecraftServer.java:1008)
[13:49:45] [Server thread/WARN]: 	at net.minecraft.server.v1_16_R3.MinecraftServer.w(MinecraftServer.java:847)
[13:49:45] [Server thread/WARN]: 	at net.minecraft.server.v1_16_R3.MinecraftServer.lambda$0(MinecraftServer.java:164)
[13:49:45] [Server thread/WARN]: 	at java.lang.Thread.run(Thread.java:748)

Hai-tun commented:
@DevSrSouza Add an executor priority fix

Hai-tun commented:
@DevSrSouza Plz, review code and merge.

Commit:
Update jetbrains exposed version

devnatan commented:
i think it would be better if you kept the `cancelOnClick` parameter of the [`menu`](https://github.com/DevSrSouza/KotlinBukkitAPI/blob/d1947bee95d078defb47659f867a73e6f759aaa9/core/src/main/kotlin/br/com/devsrsouza/kotlinbukkitapi/dsl/menu/MenuDSLImpl.kt#L35) func, for reasons of backward compatibility.

Hai-tun commented:
It seems to be able to improve compatibility by that way... but in fact, the signature of the method has changed due to the addition of parameters. So even if `cancelOnClick` is not modified, the bytecode still unable to be compatible.
